### DISCUSSION TITLE
How to use MetropolisExchange sampler in dissipative models?

### DISCUSSION THREAD

**Discussion by GillianGrayson** [netket/netket#894] [Q&A] [answered]
Hi!

Could you please provide some explanation about using `MetropolisExchange` in the dissipative systems?
Let's consider as an example MBL model, which was mentioned in #850 and [this discussion](https://github.com/netket/netket/discussions/877).
This model has symmetry: half of the spins are in the ground state, half are in the excited state.
For example, lattice size N=4. In this case, we have the following allowed states in space:
x_1 = 0011
x_2 = 0101
x_3 = 1001
x_4 = 0110
x_5 = 1010
x_6 = 1100
The number of states calculates as _n-choose-k_, where n=N, k=N/2.
As far as I understood, in the dissipative systems samples are generated in doubled Hilbert space. 
For N=4 we have (4-choose-2)^2 = 36 possible states:
x_i x_j (i, j = 1,...,6).

Which exactly graph should I provide to `MetropolisExchange`? The graph should correspond to Hilbert space or **doubled** Hilbert space?
If I use a graph, which corresponds to the doubled Hilbert space, the sampler gives all possible states, not only allowed. This is clear to me.
```python
graph = nk.graph.Hypercube(2*N, n_dim=1, pbc=False)
sa = nk.sampler.MetropolisExchange(lind.hilbert, graph=graph)
```

But if I use a graph, which corresponds to  Hilbert space (not doubled)
```python
graph = nk.graph.Hypercube(N, n_dim=1, pbc=False)
sa = nk.sampler.MetropolisExchange(lind.hilbert, graph=graph)
```
the sampler generates allowed states but **not all** (details below).

For N=4 we obtain the following clusters: [[0 1] [1 2] [2 3]]. But the sampler provides states with changed values not only on these indexes.
Example of states:
1,0,1,0,0,0,1,1
1,0,0,1,0,1,0,1
0,1,1,0,1,0,0,1
As you see, not only values at indexes 0,1,2,3 in doubled Hilbert space are changed.

The following script analyzes which exactly states generates by `MetropolisExchange` sampler (allowed or not):
```python
import netket as nk
import numpy as np
from netket.hilbert import Fock
from tqdm import tqdm


# Model params
N = 8
seed = 42
W = 1.0
U = 1.0
J = 1.0
dt = 1
gamma = 0.1

# Ansatz params
beta = 2
alpha = 2
n_samples = 10000
n_iter = 10

np.random.seed(seed)
energies = np.random.uniform(-1.0, 1.0, N)

# Hilbert space
hi = Fock(n_max=1, n_particles=N//2, N=N)

# The Hamiltonian
ha = nk.operator.LocalOperator(hi)
# List of dissipative jump operators
j_ops = []
for boson_id in range(N - 1):
    ha += W * energies[boson_id] * nk.operator.boson.number(hi, boson_id)
    ha += U * nk.operator.boson.number(hi, boson_id) * nk.operator.boson.number(hi, boson_id + 1)
    ha -= J * (nk.operator.boson.create(hi, boson_id) * nk.operator.boson.destroy(hi, boson_id + 1) + nk.operator.boson.create(hi, boson_id + 1) * nk.operator.boson.destroy(hi, boson_id))
    if dt == 1:
        A = (nk.operator.boson.create(hi, boson_id + 1) + nk.operator.boson.create(hi, boson_id)) * (nk.operator.boson.destroy(hi, boson_id + 1) - nk.operator.boson.destroy(hi, boson_id))
    elif dt == 0:
        A = nk.operator.boson.create(hi, boson_id) * nk.operator.boson.destroy(hi, boson_id)
    j_ops.append(np.sqrt(gamma) * A)
ha += W * energies[N - 1] * nk.operator.boson.number(hi, N - 1)
if dt == 0:
    A = nk.operator.boson.create(hi, N - 1) * nk.operator.boson.destroy(hi, N - 1)
    j_ops.append(np.sqrt(gamma) * A)

# Create the Liouvillian
lind = nk.operator.LocalLiouvillian(ha, j_ops)

# Neural quantum state model: Positive-Definite Neural Density Matrix using the ansatz from Torlai and Melko
ndm = nk.models.NDM(
    alpha=alpha,
    beta=beta,
)

# Metropolis Sampling
graph = nk.graph.Hypercube(N, n_dim=1, pbc=False)
sa = nk.sampler.MetropolisExchange(lind.hilbert, graph=graph)

# Variational state
vs = nk.vqs.MCMixedState(sampler=sa, model=ndm, n_samples=n_samples)
vs.init_parameters(nk.nn.initializers.normal(stddev=0.01))

# Allowed states
allowed_states = vs.hilbert.all_states()

# Allowed states statistics
states_dist = {}
for st in allowed_states:
    str_st = np.array2string(st, separator='')[1:-1].replace('.', '')
    states_dist[str_st] = 0

for it in range(n_iter):

    # Get batch of samples
    batch_of_samples = np.asarray(vs.samples.reshape((-1, vs.samples.shape[-1])))

    # Check samples
    num_non_exist = 0
    for st_id in tqdm(range(0, batch_of_samples.shape[0])):
        str_st = np.array2string(batch_of_samples[st_id, :], separator='')[1:-1].replace('.', '')
        if str_st not in states_dist:
            num_non_exist += 1
        else:
            str_st = np.array2string(batch_of_samples[st_id, :], separator='')[1:-1].replace('.', '')
            states_dist[str_st] += 1

    print(f"Number of non-existing states in space: {num_non_exist} out of {batch_of_samples.shape[0]}")

    non_zero_states = np.count_nonzero(list(states_dist.values()))
    print(f"Number of non-zero states in allowed space: {non_zero_states} out of {len(allowed_states)}")

    vs.sample()
```

I vary N at the beginning of the script:
For N=4 all allowed states are generated. No non-allowed states were generated. Everything is fine!
For N=6 only 220 out of 400 allowed states are generated (number of states = 6-choose-3 = 20). All missed states never met during multiple sampling. No non-allowed states were generated.
For N=8 only 980 out of 4900 allowed states are generated (number of states = 8-choose-4 = 70), All missed states never met during multiple sampling. No non-allowed states were generated.

I also tried different clusters for ExchangeRule (for example add all connections [[0 1] [1 2] [2 3] [0 2] [0 3] [1 3]]) - no effect.

I don't understand how to set up a graph correctly (with a sampling of **all** allowed states, without non-allowed states)  in my case.

**Comment by PhilipVinc** [comment, marked as answer]
You want to maintain the symmetry along the row and column indices independently (because it's two independent hilbert spaces). This means that the graph of possible exchanges must be the direct sum of the two graphs, which can be obtained with the disjoint union of the two.

`nk.graph.disjoint_union(physical_graph, physical_graph)`

**Reply by GillianGrayson** [reply]
Awesome! Now all sampled states are correct. Thanks a lot!

```python
graph = nk.graph.Hypercube(N, n_dim=1, pbc=False)
sa_graph = nk.graph.disjoint_union(graph, graph)
sa = nk.sampler.MetropolisExchange(lind.hilbert, graph=sa_graph)
```
